File: BOJ/10451.cpp
#include<cstdio>
using namespace std;

int ccnt;
int cycle[1001];
int arr[1001];
int check[1001];

void Reset() {
	ccnt = 0;
	for (int i = 0; i < 1001; i++) {
		arr[i] = 0;
		check[i] = 0;
		cycle[i] = 0;
	}
}

void dfs(int init) {
	if (check[init] == 1) cycle[init] = 1;
	else if (check[init] == 0) check[init] = 1;
	if (cycle[init] == 1 || check[init] == -1) return;
	dfs(arr[init]);
	check[init] = -1;
}

int main() {
	int tc;
	scanf("%d", &tc);
	while (tc--) {
		Reset();
		int N;
		scanf("%d", &N);
		for (int i = 1; i <= N; i++) {
			scanf("%d", &arr[i]);
		}

		for (int i = 1; i <= N; i++) {
			if (check[i] == 0) {
				dfs(i);
				ccnt++;
			}
		}
		printf("%d\n", ccnt);
	}

	return 0;
}

File: BOJ/python/10451.py
import sys

sys.setrecursionlimit(10 ** 7)

input = sys.stdin.readline


# 한 번 방문할 때 연결된 노드는 모두 방문
def dfs(n):
    global arr, check

    next_node = arr[n - 1]
    if not check[next_node]:  # 방문하지 않은 노드인 경우에만 방문
        check[next_node] = 1
        dfs(next_node)


def main() -> None:
    global arr, check

    # 입력
    T = int(input().rstrip('\n'))
    for _ in range(T):
        N = int(input().rstrip('\n'))
        arr = list(map(int, input().rstrip('\n').split(' ')))
        check = [0] * (len(arr) + 1)
        cnt = 0

        # 그래프 순회
        for n in arr:
            if not check[n]:  # 방문하지 않은 노드만 방문
                check[n] = 1
                dfs(n)
                cnt += 1

        print(cnt)


if __name__ == "__main__":
    main()
